### ISSUE TITLE
SIGSEGV (signal SIGSEGV: address access protected (fault address: 0x752ff03e90))

### ISSUE THREAD

**Issue by ppamorim** [realm/realm-java#6427]
#### Actual Results
App is crashing during a transaction. Returning error like `SIGSEGV (signal SIGSEGV: address access protected (fault address: 0x752ff03e90))` I bet it is a crash on the C++ core. My Realm is encrypted.
I can send a debug APK privately, will this be useful?
#### Steps & Code to Reproduce
```kotlin
val realm = Realm.getDefaultInstance()
realm.executeTransaction {
  getTask(realm, name)?.let { task ->
    task.type = newType //This works perfectly
    task.mode = null //crash here, if the remove this part of the code the crash stops.
  }
}
realm.close()
```
#### Version of Realm and tooling

Realm version(s): 5.8.0

Realm Sync feature enabled: No

Android Studio version: 3.3

Android Build Tools version: 
 - com.android.tools.build:gradle:3.3.0
 - multiDexEnabled

Gradle version: 4.10.2

Which Android version and device(s): 8.0.0 on One Plus 3T

**Comment by ppamorim** [comment]
Probably related: https://github.com/realm/realm-core/issues/2944

**Comment by taynanb** [comment]
I was able to reproduce the same problem here. As well as @ppamorim commented, the error seems to be random, occurs in different Realm method calls.

**Comment by ppamorim** [comment]
@cmelchior I think I have found the issue. I was modifying a specific object in a async thread with this same object loaded in the main thread with a change listener added. When the async thread finishes (commit the realm changes) I was calling a method on the main thread to remove and re-add the change listener, which caused the error. Removing the change listener stopped the error from happening. Shouldn't the change listener be passive?

**Comment by ppamorim** [comment]
@cmelchior I will try to replicate this on a sample app.

**Comment by cmelchior** [comment]
@ppamorim I assume you are talking about a single object listener, correct?
It should work, but given that the code around handling notifications for single objects are a bit complex I wouldn't be super surprised if we had a bug in there somewhere. Do you perhaps have a code snippet I can take a look at and see if I can reproduce it on my end?

**Comment by ppamorim** [comment]
@cmelchior Yes, it's the `RealmObjectChangeListener`.

**Comment by ppamorim** [comment]
@cmelchior Here is the example: https://github.com/ppamorim/realm-SIGSEGV-crash
Keep trying compiling it multiple times and uncommenting the lines 45-48 if the crash does not happen. I know, the example is ridiculous but it was the only way I found to simulate it on a example, my app probably was doing the same but in different threads and contexts.

**Comment by cmelchior** [comment]
@ppamorim Thank you very much for this. I can reproduce the crash.

It happens because you remove the changelistener from inside the changelistener. This should be supported but I guess there is a bug somewhere I have to track down.

If you want to work around it, for now, you can try to remove it from outside the callback, e.g. by doing it in a `runOnUiThread()`:

```
        foo?.addChangeListener<Foo> { _, _ ->
            runOnUiThread { foo?.removeAllChangeListeners() }
            foo?.addChangeListener<Foo> { _, _ -> }
        }
```

**Comment by ppamorim** [comment]
@cmelchior Thank you, I have implement states to prevent unnecessary add and remove of change listeners (I am using the `addChangeListener` with parameter, not a directly lambda on my project). I think I experienced something similar on iOS related to removing change listeners inside a change listener callback, but I am not 100% sure.

**Comment by ppamorim** [comment]
@cmelchior I agree with you, preventing any unnecessary update of the variable `foo` during the execution stops the error from happening. I simply have added a `if (foo == null)` to check if the variable is already loaded and the problem has stopped. Unfortunately this is not a solution since the reattribution of values can happen anytime for whatever reason.
